### DISCUSSION TITLE
Question about Public alternatives to Private variable/property/method usage

### DISCUSSION THREAD

**Discussion by jaredmaxwell** [sqlalchemy/sqlalchemy#12020] [Usage Questions]
We built an abstraction over SQLAlchemy to make common things less tedious. In doing so we used a bunch of private SQLAlchemy parts. I'm hoping there are public alternatives to what we used. 1.3->1.4 wasn't very easy due to our changes and we were hoping to simplify some of our abstraction where possible.

**Are there alternative public APIs to these Query private methods we use?**
- `Query`'s `_legacy_setup_joins`
- `Query`'s `_last_joined_entity`
- `Query`'s `_entity_from_pre_ent_zero`

**We put custom items in the options for some automatic delete filtering. Is there a way to save custom options during query building in the public API?**
 `_compile_options` or `_with_options`

**Using _type_affinity for type reflection is this the right way to do this with a Custom column type like our "Geometry"?**
```
@listens_for(Table, "column_reflect")
def _on_column_reflect(inspector, table, column_info):
  """
  During SQLAlchemy's reflection process, correct any datatypes that failed to resolve properly.
  https://docs.sqlalchemy.org/en/14/core/custom_types.html#working-with-custom-types-and-reflection
  """
  if column_info["type"]._type_affinity is sqltypes.NullType:
    # column type could not be determined during reflection
    from sc.common.customsql.geotypes import Geometry
    # for now, we need to update this with any new columns that have a type that cannot be detected.
    type_mapper = {
      "min_quality_geo": Geometry,
      "low_quality_geo": Geometry,
      "reg_quality_geo": Geometry,
      "high_quality_geo": Geometry,
      "max_quality_geo": Geometry,
    }
    if column_info["name"] in type_mapper:
      column_info["type"] = type_mapper[column_info["name"]]()
```

Thanks!

**Comment by zzzeek** [comment]
re: the Query stuff, not at all, you would need to illustrate use cases and we'd have to look in current patterns for what you need to do.  When something is marked as private in the API, we do that quite intentionally - it means we dont have to worry about people using it or our having to change it.   so seeking out public -API supported patterns early is always the way to go, where we can hash out what you need to do and add whatever elements might be reasonably necessary.

if you are doing delete filtering, that sounds like what we do with the do_orm_execute / with_loader_criteria pattern we document at https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#sqlalchemy.orm.with_loader_criteria and the examples at https://docs.sqlalchemy.org/en/20/orm/examples.html#module-examples.extending_query .   the [do_orm_execute](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.do_orm_execute) event hook should be your one stop for all kinds of automated query customization.

re: _type_affinity, if you need to check for nulltype you can just do `isinstance(type, NullType)`, you dont need affinity for that.  you can also get a general "affinity" version of a type calling the [as_generic](https://docs.sqlalchemy.org/en/20/core/type_api.html#sqlalchemy.types.TypeEngine.as_generic) method.
